nested-datatypes Blog Post (#40)
plclub/plclub-web: blog/2020-12-04-nested-datatypes/2020-12-04-nested-datatypes.md

---
title: Do we need nested datatypes?
author: Stephanie Weirich
tags: haskell
description: An exploration of nested types and perfect trees in Haskell.
---

> The following was generated from a [literate Haskell file](./assets/Nested.lhs).

Reviewer: ```suggestion
This blog post was generated from a [literate Haskell file](./assets/Nested.lhs).
```

Perfect trees are an example of a fairly constrained, symmetric and
artificial data structure. So was it just a fluke that we could define a
GADT analogue to the nested datatype definition?

I don't think so.

I know of at least two other examples of constrained data structures
that can be implemented by either nested datatypes and GADT-based
alternatives.

- [Well-scoped expressions](http://www.staff.city.ac.uk/~ross/papers/debruijn.html)

  A famous use of nested datatypes is to ensure that lambda calculus
  expressions are well-scoped. This idea underlies the design of Kmett's
  [bound](https://www.schoolofhaskell.com/user/edwardk/bound) library.
  
  However, instead of using a nested datatype, it is also possible to use
  a type-level natural number to bound the scopes of bound variables, as
  shown in [this
  implementation](https://github.com/sweirich/lennart-lambda/blob/master/lib/DeBruijnScoped.lhs).

- [Finger trees](http://www.staff.city.ac.uk/~ross/papers/FingerTree.html)

  Haskell's implementation of the
  [sequence](https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Sequence.html)
  data structure is built on FingerTrees. In the module

>   [DFinger.lhs](./assets/DFinger.lhs) I've sketched out a nat-indexed replacement

Reviewer 2: Better to link to the dth repo itself. That file has a link to other files in the same repo (along with a cabal file etc).

Author: Done.

  to the nested datatype.

[^0]: Many examples of nested datatypes, especially for perfect trees, use the type
  `(a,a)` instead of `Two`. However, it is convenient in modern Haskell to
   have the appropriate definitions of `fmap` etc. available for this
  auxiliary type.
[^1]: Regular datatypes produce regular trees, i.e. those that are either finite or self similar. Since the collection of perfect trees is a subset of normal binary trees (which are regular) it is a bit confusing that we use a non-regular datatype to represent them. But, note that this nested datatype definition includes a height "prefix" that prevents it from being self-similar (i.e. we cannot embed smaller values of this type directly into larger values). 
[^2]: Fritz Henglein, Type Inference with Polymorphic Recursion. ACM Transactions
  on Programming Languages and Systems. Vol 15, Issue 2. April 1993.
[^3]: Assaf J Kfoury, Jerzy  Tiuryn, Paweł Urzyczyn. Type reconstruction in the presence of polymorphic recursion. ACM Transactions on Programming Languages and Systems. Vol 15, Issue 2. April 1993.

> [^4]: https://www.haskell.org/onlinereport/decls.html#type-signatures

Reviewer: This and the URL in footnote 6 aren't clickable.
